### PR TITLE
feat: WIP multiple environment fetch api 

### LINKED ISSUE #16928: [Task]:  Create End to End flow to fetch environments & environment variables
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

- Create Domains/Models for environments & environment variables
- Create Services for the environments & environment variables
- Create controllers with endpoints.

### PR DESCRIPTION
**Pull request by sondermanish**
## Description

>  Primary commit for the multiple environment feature.

Fixes #16928 

## Type of change
- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## How Has This Been Tested?

- Not yet tested

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/acl/AclPermission.java [hunk 1]
```diff
@@ -116,7 +118,7 @@ public enum AclPermission {
     READ_PERMISSION_GROUPS("read:permissionGroups", PermissionGroup.class),
     ASSIGN_PERMISSION_GROUPS("assign:permissionGroups", PermissionGroup.class),
     UNASSIGN_PERMISSION_GROUPS("unassign:permissionGroups", PermissionGroup.class),
-
+    MANAGE_ENVIRONMENT_VARIABLE("manage:environmentVariables", EnvironmentVariable.class),
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/acl/AclPermission.java:121]: Nit: Can we follow convention here and use the plural form in the name of the permission?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/ce/EnvironmentVariableControllerCE.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+public class EnvironmentVariableControllerCE {
+
+    private EnvironmentVariableService environmentVariableService;
+
+    @Autowired
+    public EnvironmentVariableControllerCE(EnvironmentVariableService environmentVariableService) {
+        this.environmentVariableService = environmentVariableService;
+    }
+
+    @GetMapping("/{envVarId}")
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/ce/EnvironmentVariableControllerCE.java:30]: Could you please illustrate an example of needing this endpoint? Should we expose an environment level parameterized fetch all instead?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/domains/EnvironmentVariable.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import java.util.Set;
+
+@Getter
+@Setter
+@NoArgsConstructor
+@Document
+public class EnvironmentVariable extends BaseDomain {
+
+
+    String environmentId;
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/domains/EnvironmentVariable.java:21]: Not sure I see what this refers to?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/domains/EnvironmentVariable.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+    String applicationId;
+
+    String workspaceId;
+
+    String name;
+
+    //subject to type change;
+    @JsonProperty
+    String value;
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/domains/EnvironmentVariable.java:31]: Let's add logic for hidden and encryption here as well

File: app/server/appsmith-server/src/main/java/com/appsmith/server/dtos/EnvironmentVariableDTO.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+import java.time.Instant;
+import java.util.HashSet;
+import java.util.Set;
+
+@Getter
+@Setter
+@NoArgsConstructor
+@ToString
+@Slf4j
+public class EnvironmentVariableDTO {
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/dtos/EnvironmentVariableDTO.java:26]: What's the use case for this class?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/repositories/ce/CustomEnvironmentVariableRepositoryCE.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import java.util.List;
+
+public interface CustomEnvironmentVariableRepositoryCE extends AppsmithRepository<EnvironmentVariable> {
+
+    Mono<EnvironmentVariable> findById(String id, AclPermission aclPermission);
+
+    Flux<EnvironmentVariable> findByApplicationId(String id, AclPermission aclPermission);
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/repositories/ce/CustomEnvironmentVariableRepositoryCE.java:15]: Let's only add the methods that we hope to use in this iteration?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/EnvironmentVariableServiceCEImpl.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+import com.appsmith.server.domains.EnvironmentVariable;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.mongodb.core.ReactiveMongoTemplate;
+import org.springframework.data.mongodb.core.convert.MongoConverter;
+import org.springframework.util.MultiValueMap;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Scheduler;
+import com.appsmith.server.services.*;
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/EnvironmentVariableServiceCEImpl.java:13]: Nit: Please use specific import statements

Comment (sondermanish) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/EnvironmentVariableServiceCEImpl.java:13]: Have removed the wild card imports.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/ce/EnvironmentVariableControllerCE.java (hunk 2 above)
Comment (sondermanish) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/controllers/ce/EnvironmentVariableControllerCE.java:30]: We would ideally never use this endpoint to fetch individual variables, kept it for check, will replace it with the one at environment level,

File: app/server/appsmith-server/src/main/java/com/appsmith/server/domains/EnvironmentVariable.java (hunk 3 above)
Comment (sondermanish) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/domains/EnvironmentVariable.java:21]: `environmentId` Is kept to associate with the environment to which they belong.

**Comment by sondermanish** [comment]
The PR is moved to Enterprise edition, hence closing this stale draft PR

**Pull request closed without merging**